exercises/src/main/scala/fpinscala/datastructures/List.scala

```diff
@@ -50,19 +52,112 @@ object List { // `List` companion object. Contains functions for creating and wo
     foldRight(ns, 1.0)(_ * _) // `_ * _` is more concise notation for `(x,y) => x * y`; see sidebar
 
 
-  def tail[A](l: List[A]): List[A] = ???
+  def tail[A](l: List[A]): List[A] = l match {
+    case Nil => throw new Exception
+    case Cons(_, t) => t
+  }
+
+  def setHead[A](l: List[A], h: A): List[A] = l match {
+    case Nil => Cons(h, Nil)
+    case Cons(_, t) => Cons(h, t)
+  }
+
+  def drop[A](l: List[A], n: Int): List[A] = {
+    if (n <= 0) l
+    else l match {
+      case Nil => Nil
+      case Cons(_, t) => drop(t, n - 1)
+    }
+  }
+
+  def dropWhile[A](l: List[A], f: A => Boolean): List[A] = l match {
+    case Nil => Nil
+    case Cons(h, t) =>
+      if (f(h)) dropWhile(t, f)
+      else l
+  }
+
+  def init[A](l: List[A]): List[A] = l match {
+      case Nil => Nil
+      case Cons(_, Nil) => Nil
+      case Cons(h, t) => Cons(h, init(t))
+  }
+
+  def length[A](l: List[A]): Int = l match {
+    case Nil => 0
+    case Cons(_, Nil) => 1
+    case Cons(_, t) => 1 + length(t)
+  }
+
+  //length with foldRight
+  def length2[A](l: List[A]): Int = foldRight(l, 0)((_: A, b: Int) => b + 1)
+
+  @tailrec
+  def foldLeft[A,B](l: List[A], z: B)(f: (A, B) => B): B = l match {
+    case Nil => z
+    case Cons(h, t) => foldLeft(t, f(h, z))(f)
+  }
 
-  def setHead[A](l: List[A], h: A): List[A] = ???
+  def sumFL(ns: List[Int]) =
+    foldLeft(ns, 0)((x,y) => x + y)
 
-  def drop[A](l: List[A], n: Int): List[A] = ???
+  def productFL(ns: List[Double]) =
+    foldLeft(ns, 1.0)(_ * _)
 
-  def dropWhile[A](l: List[A], f: A => Boolean): List[A] = ???
+  def reverse[A](l: List[A]): List[A] = foldLeft(l, List[A]())((y, x) => Cons(y, x))
```

Comment:
nit: Having `(y, x)` instead of `(x, y)` is confusing. It can also be simplified to:

```
foldLeft(l, List[A]())(Cons)
```